File: src/lss/js/code_parser.ts
import { parse } from "@babel/parser";
import traverse, { NodePath } from "@babel/traverse";
import * as t from "@babel/types";

interface NodeMeta {
    type: "function" | "class";
    source: string;
    start_line: number;
    end_line: number;
    start_column: number;
    end_column: number;
}

/**
 * Gets the "name" of a function or class node, even if it's an anonymous
 * expression assigned to a variable or a class method.
 */
function getNodeName(path: NodePath): string {
    const { node, parent } = path;

    // Handles `function MyFunc() {}` and `class MyClass {}`
    if ("id" in node && node.id) {
        return node.id.name;
    }

    // Handles `const MyFunc = () => {}`
    if (t.isVariableDeclarator(parent) && t.isIdentifier(parent.id)) {
        return parent.id.name;
    }

    // Handles `class A { myMethod() {} }` and `const obj = { myMethod() {} }`
    if (t.isClassMethod(node) || t.isObjectMethod(node)) {
        if (t.isIdentifier(node.key)) {
            return node.key.name;
        }
    }

    return "(anonymous)";
}

/**
 * A direct TypeScript/Babel equivalent of the provided Python AST parser.
 */
export function parseCode(code: string): Record<string, NodeMeta> {
    const structure: Record<string, NodeMeta> = {};
    if (!code.trim()) return structure;

    const ast = parse(code, {
        sourceType: "module",
        plugins: ["typescript", "jsx"],
        errorRecovery: true,
    });

    /**
     * This is the equivalent of the Python `visit` function. It's a visitor
     * that will be applied recursively.
     */
    const visitor = {
        "FunctionDeclaration|FunctionExpression|ArrowFunctionExpression|ClassDeclaration|ClassExpression|ClassMethod|ObjectMethod"(
            path: NodePath<any>
        ) {
            const name = getNodeName(path);
            const isClass = path.isClass() || path.isClassExpression();

            // Equivalent to `parents + [child.name]`
            const qualnameParts = [...(this.parents as string[]), name];
            const qualname = qualnameParts.join(".");

            const { node } = path;
            const { start, end, loc } = node;

            structure[qualname] = {
                type: isClass ? "class" : "function",
                source: code.slice(start ?? 0, end ?? 0),
                start_line: loc?.start.line ?? 0,
                end_line: loc?.end.line ?? 0,
                start_column: loc?.start.column ?? 0,
                end_column: loc?.end.column ?? 0,
            };

            // Recurse into children with the updated parent context.
            // This is the direct equivalent of `visit(child, qualname_parts)`.
            path.traverse(visitor, { parents: qualnameParts });

            // We handled the children manually, so skip Babel's default traversal.
            path.skip();
        },
    };

    // Start the traversal with an empty parent list.
    // This is the equivalent of the initial `visit(tree, [])` call.
    traverse(ast, visitor, undefined, { parents: [] });

    return structure;
}

File: src/lss/py/code_parser.py
import ast
from typing import Dict


def parse_code(code: str) -> Dict[str, dict]:
    """
    Parse Python source code into an AST and extract the source text of
    all functions and classes (including nested ones), along with their types.

    The result mirrors `backend.core.diff_parser.parse_code_structure` and is
    a mapping of qualified names to metadata, for example:

    - ``foo``          – top‑level function
    - ``Foo``          – top‑level class
    - ``Foo.bar``      – method ``bar`` inside class ``Foo``
    - ``main.inner``   – nested function ``inner`` inside ``main``
    """
    structure: Dict[str, dict] = {}
    if not code:
        return structure

    try:
        tree = ast.parse(code)

        def visit(node: ast.AST, parents: list[str]) -> None:
            for child in ast.iter_child_nodes(node):
                if isinstance(
                    child,
                    (ast.FunctionDef, ast.AsyncFunctionDef, ast.ClassDef),
                ):
                    node_type = "class" if isinstance(
                        child, ast.ClassDef
                    ) else "function"

                    qualname_parts = parents + [child.name]
                    qualname = ".".join(qualname_parts)

                    structure[qualname] = {
                        "type": node_type,
                        "source": ast.get_source_segment(code, child),
                        "start_line": getattr(child, "lineno", 0),
                        "end_line": getattr(
                            child, "end_lineno", getattr(child, "lineno", 0)
                        ),
                        "start_column": getattr(child, "col_offset", 0),
                        "end_column": getattr(
                            child,
                            "end_col_offset",
                            getattr(child, "col_offset", 0),
                        ),
                    }

                    visit(child, qualname_parts)
                else:
                    visit(child, parents)

        visit(tree, [])

    except (SyntaxError, TypeError):
        # Gracefully handle files that are not valid Python
        pass

    return structure
